### PR TITLE
Add currrent work of concurrency for DMSDK for suggestions ##1269

### PR DESCRIPTION
**Pull request by llinggit**
So we can incorporate your pull request, please share the following:
* What issue are you addressing with this pull request?
* Are you modifying the correct branch? (See CONTRIBUTING.md)
* Have you run unit tests? (See CONTRIBUTING.md)
* Version of MarkLogic Java Client API (see Readme.txt)
* Version of MarkLogic Server (see admin gui on port 8001)
* Java version (`java -version`)
* OS and version
* What Changed: What happened before this change? What happens without this change?

### CODE REVIEW

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java [hunk 1]
```diff
@@ -272,32 +272,62 @@
...
   public QueryBatcher withBatchSize(int docBatchSize, int docToUriBatchRatio);
 
+  /**
+   * Sets the number of uris to process per batch, the ratio of uris-to-retrieve/uris-to-process and threadThrottleFactor.
+   * If not set, the default docBatchSize and docToUriBatchRatio are calculated and set according to server status. The
+   * default threadThrottleFactor is 0, which means no throttle at all. For example, if docBatchSize is 100 and
+   * docToUriBatchRatio is 5, there is one thread to retrieve 500 uris and process 100 uris and there are 4 threads to
+   * process the other 400 uris.
+   * @return this instance for method chaining
+   */
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java:300]: Maybe this is the overload where the JavaDoc should add a description of how threading works?  Possibly something like

Ordinarily, QueryBatcher starts a separate thread for each document processing batch (and also collects document uris in one of these threads). The threadThrottleFactor reduces the number of default threads. Thus, a threadThrottleFactor equal to the docToUriBatchRatio  - 1 runs in a single thread. A threadThrottleFactor of 0 explicitly uses the maximum number of threads.

**Review by ehennum — APPROVED** [review]
Thanks for the scrupulous and intelligent work on increasing the level of concurrent processing.

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java [hunk 2]
```diff
@@ -272,32 +272,62 @@
...
 
   /**
-   * Sets the number of uris to retrieve per batch. Since uris are small
-   * relative to full documents, this number should be much higher than the
-   * batch size for WriteBatcher. The default batch size is 1000.
-   *
+   * Before release 5.3.2, this method sets the number of uris to retrieve per batch. Since uris are small relative to full
+   * documents, this number should be much higher than the batch size for WriteBatcher. The default batch size is 1000.
+   * After release 5.3.2, this method sets the number of uris to process per batch. If not set, the default docBatchSize
+   * is calculated and set according to server status.
```
[the remark this answers is not in the document]
Comment (llinggit) [reply marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java:278]: simplified the JavaDoc and added missing parameter.

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java [hunk 3]
```diff
@@ -272,32 +272,62 @@
...
   public QueryBatcher withBatchSize(int docBatchSize);
 
-  // new setter methods
+  /**
+   * Sets the number of uris to process per batch and the ratio of uris-to-retrieve/uris-to-process. If not set,
+   * the default docBatchSize and docToUriBatchRatio are calculated and set according to server status. For example,
+   * if docBatchSize is 100 and docToUriBatchRatio is 5, there is one thread to retrieve 500 uris and process 100 uris
+   * and there are 4 threads to process the other 400 uris.
+   * @return this instance for method chaining
+   */
```
[the remark this answers is not in the document]
Comment (llinggit) [reply marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java:290]: Thanks for the suggestions. Modified the JavaDoc.

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java (hunk 1 above)
Comment (llinggit) [reply marklogic-client-api/src/main/java/com/marklogic/client/datamovement/QueryBatcher.java:300]: Thanks for the suggestions. Modified the JavaDoc.

**Pull request merged**